Cleanup lettuce instrumentation and add tests

twcrone opened the pull request:
Provides Java instrumentation for the Lettuce driver for Redis. Instrumentation is being migrated from New Relic Experimental to be included in the product

twcrone commented on instrumentation/lettuce-5.0/src/test/java/com/nr/lettuce5/instrumentation/Lettuce5InstrumentationTest.java:
```diff
@@ -0,0 +1,170 @@
+        // and required datastore metrics are sent
+        DatastoreHelper helper = new DatastoreHelper(DatastoreVendor.Redis.name());
+        helper.assertAggregateMetrics();
+        // FIXME: Why is this 2 when 4.3 and 6 return 1?
```
No idea why for 5, we seem to get double DB metrics.  6 and 4.3 with same test, only show 1

twcrone commented on build.gradle:
```diff
@@ -7,6 +7,7 @@ plugins {
 
 allprojects {
     apply plugin: 'idea'
+    apply plugin: 'eclipse'
```
Doug added this and I was wondering if we should remove or leave for eclipse users.

jasonjkeller replied:
I would say delete it, but I don't think it particularly matters.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRErrorConsumer.java:
```diff
@@ -0,0 +1,24 @@
+	@Override
+	public void accept(Throwable t) {
+		NewRelic.noticeError(t);
+		if(holder != null && !holder.hasEnded()) {
```
Doesn't look formatted.

twcrone replied:
yeah, I think I need to format 4.3 and 5 source

jasonjkeller commented on instrumentation/lettuce-4.3/build.gradle:
```diff
@@ -0,0 +1,24 @@
+    implementation(project(":newrelic-weaver-api"))
+    implementation(project(":agent-bridge"))
+    implementation 'biz.paluch.redis:lettuce:4.4.1.Final'
+    testImplementation('org.testcontainers:testcontainers:1.17.1')
```
Are we requiring Docker to run non-AIT agent tests now? If so this needs to be stated as a requirement in the agent readme.

jasonjkeller commented on instrumentation/lettuce-4.3/build.gradle:
```diff
@@ -0,0 +1,24 @@
+}
+
+verifyInstrumentation {
+    passes 'biz.paluch.redis:lettuce:[4.3.0,4.4.4.Final]'
```
It's a best practice to use `passesOnly` when you can as it will not only verify the supported version defined in the range but also verify that all versions outside of the range don't verify. Doing this helps to prevent scenarios where multiple instrumentation modules apply and cause double instrumentation.

Also, has the verifier actually been run against this?

twcrone replied:
true, need to do that still...

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/lambdaworks/redis/AbstractRedisAsyncCommands.java:
```diff
@@ -0,0 +1,55 @@
+		}
+		DatastoreParameters params = null;
+		if (uri != null) {
+
```
Delete extra line.

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/nr/fit/lettuce/instrumentation/NRBiConsumer.java:
```diff
@@ -0,0 +1,38 @@
+package com.nr.fit.lettuce.instrumentation;
```
We shouldn't have any packages containing `fit` (aka Field Instrumentation Team) if we're porting this from their experimental repo to the agent proper.

twcrone replied:
ok

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/lambdaworks/redis/RedisClient_instrumentation.java:
```diff
@@ -0,0 +1,29 @@
+
+@SuppressWarnings("deprecation")
+@Weave(originalName="com.lambdaworks.redis.RedisClient")
+public abstract class RedisClient_instrumentation extends AbstractRedisClient {
```
Should be a capital `I` for instrumentation to keep it consistent with other modules: `RedisClient_Instrumentation`

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/lambdaworks/redis/StatefulRedisConnectionImpl.java:
```diff
@@ -0,0 +1,22 @@
+import com.newrelic.api.agent.weaver.Weave;
+
+@Weave(type = MatchType.BaseClass)
+public abstract class StatefulRedisConnectionImpl<K, V> implements StatefulConnection<K, V> {
```
Best practice is to name any instrumented library classes with `_Instrumentation` appended to the original class name (e.g. `StatefulRedisConnectionImpl_Instrumentation<K, V>`) and then specify the original class name using `@Weave(type = MatchType.BaseClass, originalName="com.lambdaworks.redis.StatefulRedisConnectionImpl<K, V>")`.

This was done correctly for `RedisClient_Instrumentation` and we should be consistent about that.

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/nr/fit/lettuce/instrumentation/StringUtils.java:
```diff
@@ -0,0 +1,44 @@
+			}
+		}
+		return sb.toString();
+
```
extra space

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/nr/fit/lettuce/instrumentation/StringUtils.java:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	public static String getCSSFromArray(Object[] objects) {
+		if (objects == null || objects.length == 0)
```
Please never use this pattern of omitting the `{}` for an if statement. It's incredibly easy for this to turn into a subtle and difficult to troubleshoot bug if someone decides that they're going to add more lines to the conditional logic. It's well worth being explicit with the parens.

jasonjkeller commented on instrumentation/lettuce-5.0/build.gradle:
```diff
@@ -0,0 +1,22 @@
+}
+
+verifyInstrumentation {
+    passes 'io.lettuce:lettuce-core:[5.0.0.RELEASE,6.0.0.RELEASE)'
```
Same comment about using `passesOnly`

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRBiConsumer.java:
```diff
@@ -0,0 +1,38 @@
+package com.nr.fit.lettuce.instrumentation;
```
Same comment about `fit` in package name

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRHolder.java:
```diff
@@ -0,0 +1,69 @@
+import com.newrelic.api.agent.Token;
+
+public class NRHolder
+{
```
Formatting. This parenthesis usage is inconsistent with the rest of the codebase.

jasonjkeller commented on instrumentation/lettuce-4.3/src/main/java/com/lambdaworks/redis/AbstractRedisAsyncCommands.java:
```diff
@@ -0,0 +1,55 @@
+import com.nr.fit.lettuce.instrumentation.NRBiConsumer;
+
+@Weave(type = MatchType.BaseClass)
+public abstract class AbstractRedisAsyncCommands<K, V> {
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRHolder.java:
```diff
@@ -0,0 +1,69 @@
+    else
+    {
+        String operation = params.getOperation();
+        if(!operation.equalsIgnoreCase("expire")) {
```
Formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRSignalTypeConsumer.java:
```diff
@@ -0,0 +1,22 @@
+
+	@Override
+	public void accept(SignalType t) {
+		if(holder != null && !holder.hasEnded()) {
```
formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRSubscribeConsumer.java:
```diff
@@ -0,0 +1,22 @@
+
+	@Override
+	public void accept(Subscription t) {
+		if(holder != null) {
```
formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/Utils.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Utils {
+
+	
```
extra spaces

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/com/nr/fit/lettuce/instrumentation/Utils.java:
```diff
@@ -0,0 +1,16 @@
+	public static boolean initialized = false;
+
+	public static void init() {
+		if(instance == null) {
```
formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/AbstractRedisAsyncCommands.java:
```diff
@@ -0,0 +1,58 @@
+import io.lettuce.core.protocol.RedisCommand;
+
+@Weave
+public abstract class AbstractRedisAsyncCommands<K, V> {
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/AbstractRedisReactiveCommands.java:
```diff
@@ -0,0 +1,103 @@
+import reactor.core.publisher.SignalType;
+
+@Weave
+public abstract class AbstractRedisReactiveCommands<K, V>
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/AbstractRedisReactiveCommands.java:
```diff
@@ -0,0 +1,103 @@
+	private final StatefulConnection<K, V> connection = Weaver.callOriginal();
+
+	public <T> Mono<T> createMono(Supplier<RedisCommand<K, V, T>> commandSupplier)
+	{
```
Inconsistent paren use

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/RedisClient.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.Duration;
+
+@Weave(originalName = "io.lettuce.core.RedisClient")
+public abstract class RedisClient extends AbstractRedisClient {
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java:
```diff
@@ -0,0 +1,21 @@
+import io.lettuce.core.codec.RedisCodec;
+
+@Weave(type = MatchType.BaseClass)
+public abstract class StatefulRedisConnectionImpl<K, V> implements StatefulConnection<K, V> {
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/protocol/CommandHandler.java:
```diff
@@ -0,0 +1,30 @@
+import com.newrelic.api.agent.weaver.Weaver;
+
+import io.netty.channel.ChannelHandlerContext_Instrumentation;
+//import io.netty.channel.ChannelHandlerContext_Instrumentation;
```
Delete commented out code.

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/protocol/CommandHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+	@Trace(async=true)
+	public void channelRead(ChannelHandlerContext_Instrumentation ctx, Object msg) {
+		if(ctx.pipeline().lettuceLayerToken != null) {
```
formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/protocol/CommandHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+	@Trace(async=true)
+	public void write(ChannelHandlerContext_Instrumentation ctx, Object msg, ChannelPromise promise) {
+		if(ctx.pipeline().lettuceLayerToken != null) {
```
formatting

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/lettuce/core/protocol/CommandHandler.java:
```diff
@@ -0,0 +1,30 @@
+import io.netty.channel.ChannelPromise;
+
+@Weave
+public abstract class CommandHandler {
```
Same comment about naming weaved library classed with `_Instrumentation` and specifying the `originalName` in the `Weave` annotation.

twcrone replied:
ok, did this in 6 but didn't do in older versions...will do

jasonjkeller commented on instrumentation/lettuce-5.0/src/main/java/io/netty/channel/AbstractChannel_Instrumentation.java:
```diff
@@ -0,0 +1,65 @@
+@Weave(type=MatchType.BaseClass,originalName="io.netty.channel.AbstractChannel")
+public abstract class AbstractChannel_Instrumentation  {
+
+	public abstract ChannelPipeline_Instrumentation pipeline();
```
weird indentation

jasonjkeller commented on instrumentation/lettuce-6.0/build.gradle:
```diff
@@ -0,0 +1,24 @@
+}
+
+verifyInstrumentation {
+    passes 'io.lettuce:lettuce-core:[6.0.0.RELEASE,)'
```
Same comment about using `passesOnly`

twcrone replied:
👍

jasonjkeller commented on instrumentation/lettuce-6.0/src/main/java/com/nr/fit/lettuce/instrumentation/NRBiConsumer.java:
```diff
@@ -0,0 +1,38 @@
+package com.nr.fit.lettuce.instrumentation;
```
Same comment about `fit` in package name

jasonjkeller reviewed (changes requested):
Left some specific comments but there's also a few general things to address.

All source code files need the [New Relic copyright header](https://github.com/newrelic/newrelic-java-agent/blob/4e5074c3f91a300d98989bcf6103b70de0936da9/newrelic-agent/src/main/java/com/newrelic/agent/Transaction.java#L1-L6) but modified for 2022.

What’s up with the `.cache-tests` files being checked in? Can we delete them from this PR?
- `instrumentation/scala-2.12.0/.cache-tests`
- `instrumentation/slick-2.12_3.2.0/.cache-tests`
- `instrumentation/akka-http-core-10.2.0/.cache-tests`

There's some inconsistent formatting throughout, make sure that all files have been formatted using the [java-agent-code-style](https://github.com/newrelic/newrelic-java-agent/blob/main/dev-tools/code-style/java-agent-code-style.xml).

twcrone commented:
Looks like I have some work.  I'll try to get started this weekend in my "copious free time" but might have to pick up first thing Monday.

twcrone commented:
@jasonjkeller I think I've addressed your concerns.  I'm wondering if you wanted more in the README.md about Testcontainers?  Right now, I think only kafka and lettuce use them for instrumentation tests, so just noted that Docker might be needed to run instrumentation tests with links to Testcontainer documentation and Docker install.  Let me know your thoughts!

jasonjkeller commented:
> @jasonjkeller I think I've addressed your concerns. I'm wondering if you wanted more in the README.md about Testcontainers? Right now, I think only kafka and lettuce use them for instrumentation tests, so just noted that Docker might be needed to run instrumentation tests with links to Testcontainer documentation and Docker install. Let me know your thoughts!

Yea, just making it clear that Docker is a requirement to run all tests is what I was getting at.

jasonjkeller commented on instrumentation/lettuce-4.3/src/test/java/com/nr/lettuce43/instrumentation/helper/Data.java:
```diff
@@ -0,0 +1,11 @@
+package com.nr.lettuce43.instrumentation.helper;
```
Test source files require the copyright header as well.

twcrone replied:
oh, ok...didn't know test files too...will do!

jasonjkeller commented on instrumentation/lettuce-4.3/src/test/java/com/nr/lettuce43/instrumentation/helper/RedisDataService.java:
```diff
@@ -0,0 +1,54 @@
+package com.nr.lettuce43.instrumentation.helper;
```
Add copyright header.

jasonjkeller commented on instrumentation/lettuce-5.0/src/test/java/com/nr/lettuce5/instrumentation/helper/RedisDataService.java:
```diff
@@ -0,0 +1,81 @@
+package com.nr.lettuce5.instrumentation.helper;
```
Add copyright header.

jasonjkeller commented on instrumentation/lettuce-6.0/src/main/java/io/lettuce/core/protocol/CommandHandler_Instrumentation.java:
```diff
@@ -0,0 +1,28 @@
+package io.lettuce.core.protocol;
```
Add copyright header.

jasonjkeller commented on instrumentation/lettuce-6.0/src/test/java/com/nr/lettuce6/instrumentation/helper/RedisDataService.java:
```diff
@@ -0,0 +1,81 @@
+package com.nr.lettuce6.instrumentation.helper;
```
Add copyright header.

jasonjkeller reviewed (approved):
Just make sure that all Java class files (including tests) contain the copyright header. Otherwise looks good! Thanks for cleaning it up!
